### DISCUSSION TITLE
When a `StreamProvider` is being rebuilt, how to reuse the last value of the previous stream and skip the loading initial value?

### DISCUSSION THREAD

**Discussion by ValentinVignal** [rrousselGit/riverpod#1703] [Q&A] [answered]
I have this code snippet (you can also checkout https://github.com/ValentinVignal/flutter_app_stable/tree/riverpod/combine-streams)

```dart
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

void main() {
  runApp(const MyApp());
}

final searchProvider = StateProvider.autoDispose<String>((ref) => '');

Stream<String> getStream(String search) async* {
  await Future.delayed(const Duration(milliseconds: 500));
  yield search;
}

final streamProvider = StreamProvider.autoDispose<String>((ref) {
  final search = ref.watch(
    searchProvider,
  ); // Simulates that the stream will be recomputed every time the search changes.
  return getStream(search);
});

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return ProviderScope(
      child: MaterialApp(
        home: Scaffold(
          appBar: AppBar(
            title: const SearchTextField(),
          ),
          body: const Screen(),
        ),
      ),
    );
  }
}

class SearchTextField extends ConsumerWidget {
  const SearchTextField({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    return TextField(
      onChanged: (value) {
        ref.read(searchProvider.notifier).state = value;
      },
    );
  }
}

class Screen extends ConsumerWidget {
  const Screen({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final searchInStream = ref.watch(streamProvider);
    return Center(
      child: searchInStream.when(
        data: (data) => Text(data),
        loading: () => const CircularProgressIndicator(),
        error: (error, stack) => Text(error.toString()),
      ),
    );
  }
}
```

Every time the user types a letter in the search bar, the `streamProvider` will rebuilt and the value gotten by `ref.watch(streamProvider)` will be a `AsyncLoading`.
Is there a way to get the last value of the previous stream instead of the initial `AsyncLoading`?

What is happening now:
1. It starts the app
2. It gets in the initial loading state waiting for the first value of the stream and showing the `CircularProgressIndicator` (this one is fine).
3. The stream has emitted the initial value `''`, it reloads the UI and the `CircularProgressIndicator` disappears.
4. The user types a letter (ex: `'a'`) in the search bar.
5. The `streamBuilder` rebuilds with a new stream and `ref.watch(streamProvider)`returns `AsyncLoading` while waiting for the first value of the stream. The UI shows the `CircularProgressIndicator`. _(This is what I want to avoid)_
6. The new stream has emitted the value `'a'`, it reloads the UI and it is displayed on the screen.
7. The user types a letter (ex: `'b'`) in the search bar.
5. The `streamBuilder` rebuilds with a new stream and `ref.watch(streamProvider)`returns `AsyncLoading` while waiting for the first value of the stream. The UI shows the `CircularProgressIndicator`. _(This is what I want to avoid)_
6. The new stream has emitted the value `'ab'`, it reloads the UI and it is displayed on the screen.

Here is a video illustrating it:

https://user-images.githubusercontent.com/32538273/192678207-ca6d112a-af33-44c6-94e8-9329626e7064.mov

What I would want to do:
1. It starts the app
2. It gets in the initial loading state waiting for the first value of the stream and showing the `CircularProgressIndicator` (this one is fine).
3. The stream has emitted the initial value `''`, it should reload and the `CircularProgressIndicator` should disappear.
4. The user types a letter (ex: `'a'`) in the search bar.
5. The `streamBuilder` rebuilds with a new stream. But `ref.watch(streamProvider)` returns `AsyncData('')` (last value of the previous stream) instead of `AsyncLoading` while waiting for the first value of the stream. There is no `CircularProgressIndicator` in the UI. _(This is where the behavior changes)_
6. The new stream has emitted the value `'a'`, it reloads the UI and it is displayed on the screen.
8. The user types a letter (ex: `'b'`) in the search bar.
5. The `streamBuilder` rebuilds with a new stream But `ref.watch(streamProvider)` returns `AsyncData('a')` (last value of the previous stream) instead of `AsyncLoading` while waiting for the first value of the stream. The UI shows `'a'`. _(This is where the behavior changes)_
6. The new stream has emitted the value `'ab'`, it reloads the UI and it is displayed on the screen.

I'm ***NOT*** trying to debounce, I'm fine with creating a new stream every time the user types a letter.
What I'm trying to do is to skip the `AsyncLoading` and reuse the previous state when available.

---

How can I achieve that?

**Comment by TimWhiting** [comment]
Version 2 of riverpod does what you want. You can try out the dev release if you want, which is fairly stable, or you can wait another week or two for the stable release.

**Reply by rrousselGit** [reply]
Well about that...

The change has been reverted to be unique to `ref.refresh`/`invalidate`
If a change is triggered by `ref.watch`, the provider goes to loading again.

**Reply by rrousselGit** [reply]
You could have that behavior by doing:

```dart
ref.listen(provider, (prev, next) {
  ref.invalidateSelf();
});
```

**Comment by kentnek** [comment]
@rrousselGit in version `2.1.1` there is no `invalidateSelf` method in `ref`. How can I solve this?

**Reply by kentnek** [reply, marked as answer]
oh never mind I just found out about `when.skipLoadingOnReload`
